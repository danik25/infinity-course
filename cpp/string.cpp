/****************************************************************
*                              String
* 						=====================
* File Name: string.cpp
* Related files: string.hpp string_test.cpp
* #Version: V 1.0
* Writer: Dani Benjamin      
* Reviewer: Ayala Rosel    
* Last update: 23/07/19
****************************************************************/
#include <cstring> /* strlen */
#include <iostream> /* cout */
#include "string.hpp"

namespace hrd9
{
	/* constructor */
	String::String(const char* str)
	{
		size_t len = strlen(str);
		m_str = new char [len + 1];
		strcpy(m_str, str);
	}

	/* distructor */
	String::~String()
	{
		delete [](m_str);
	}

	/* copy constructor */
	String::String(const String& other) : m_str(init_new_str(other.m_str))
	{
		;
	}

	/* copy assignment operator */
	String& String :: operator=(const String& other)
	{
		delete[] m_str;
		m_str = init_new_str(other.m_str);

		return(*this);
	}

/*===================== non member functions =====================*/

/*===================== member functions =====================*/

	String& String::operator+=(const String& other)
	{
		size_t total_len = strlen(m_str) + strlen(other.m_str);

		String tmp = m_str;
		delete [] m_str;

		m_str = new char [total_len + 1];
		char * runner = m_str;

		strcpy(runner, tmp.m_str);
		strcat(runner, other.m_str);

		return(*this);
	}

	size_t String::length() const
	{
		return(strlen(m_str));
	}

	const char& String::operator[](size_t index) const
	{
		return(m_str[index]);
	}

	char& String::operator[](size_t index)
	{
		return(m_str[index]);
	}

	const char* String::to_c_str() const
	{
		return(m_str);
	}

/*===================== friend functions =====================*/

	const String operator+(const String& lhs, const String& rhs)
	{
		String ans = lhs;
		ans += rhs;

		return(ans);
	}

	bool operator==(const String& lhs, const String& rhs)
	{
		return(!strcmp(lhs.m_str, rhs.m_str));
	}

	bool operator!=(const String& lhs, const String& rhs)
	{
		return(strcmp(lhs.m_str, rhs.m_str));
	}

	bool operator>(const String& lhs, const String& rhs)
	{
		return(strcmp(lhs.m_str, rhs.m_str));
	}

	bool operator<(const String& lhs, const String& rhs)
	{
		return(strcmp(rhs.m_str, lhs.m_str));
	}

	std::ostream& operator<<(std::ostream& os, const String& str)
	{
		os << str.m_str;
		return(os);
	}

/*===================== assistance functions =====================*/

	char* String::init_new_str(const char *str)
	{
		size_t len = strlen(str);
		char *new_str = new char [len + 1];
		strcpy(new_str, str);

		return(new_str);
	}
}