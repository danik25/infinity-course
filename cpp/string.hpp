/****************************************************************
*                              String
* 						=====================
* File Name: string.hpp
* Related files: string.cpp string_test.cpp
* #Version: V 1.0
* Writer: Dani Benjamin      
* Reviewer: Ayala Rosel    
* Last update: 23/07/19
****************************************************************/
#ifndef ILHRD_STRING_HPP
#define ILHRD_STRING_HPP

#include <iosfwd> //ostream forward declarations

namespace hrd9
{

class String;
//non member functions:
const String operator+(const String& lhs, const String& rhs);
bool operator==(const String& lhs, const String& rhs);
bool operator!=(const String& lhs, const String& rhs);
bool operator>(const String& lhs, const String& rhs);
bool operator<(const String& lhs, const String& rhs);
std::ostream& operator<<(std::ostream& os, const String& str);
std::istream& operator>>(std::istream& is, String& str);

class String
{

public:
	// explicitly non explicit constructor
	String(const char* str = "");
	String(const String& other);
	String& operator=(const String& other);
	~String();

	String& operator+=(const String& other);
	size_t length() const;

	const char& operator[](size_t index) const; //read string[]
	char& operator[](size_t index);             //string[i] assignment

	const char* to_c_str() const;

private:
	friend bool operator==(const String& lhs, const String& rhs);
	friend bool operator!=(const String& lhs, const String& rhs);
	friend bool operator>(const String& lhs, const String& rhs);
	friend bool operator<(const String& lhs, const String& rhs);
	friend std::ostream& operator<<(std::ostream& os, const String& str);
	friend std::istream& operator>>(std::istream& is, String& str);

	char* init_new_str(const char *str);

	char* m_str;
};
} //namespace hrd9




#endif /* ILHRD_STRING_HPP */