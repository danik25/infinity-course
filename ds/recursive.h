
#ifndef  _REC_H_
#define  _REC_H_
#include <stddef.h>
/****************************************************************
*                                 Recursion 
*                       ==============================
* File Name: recursive.h
* Related files: recursive.c, recursive_test.c
* #Version: V 1.8
* Writer: Dani Benjamin                       
* Reviewer: Ayala Rosel 
* Last update: 28/05/19
****************************************************************/

/*
 * API Design: 
 * An introduction to recursion, that consists of a calculation fibonacci
 * sequence sum with iterative and recursive solution. implementation of a few
 * string.h library functions using recursion.
 */ 

#ifndef _SL_LIST_
typedef struct sl_node
 {
                    
        void * data;
        struct sl_node *next; 

} sl_node_t;
#endif /* _SL_LIST_ */

/* Description:
 * Recives an index in a fibonacci sequence and returns its value using
 * recursion. The sequence starts 0 1 1 2...
 * 
 *
 * Params:
 *        index - an index in a fibonacci sequence
 *
 * Return Value: value of a givin index. 
 */
size_t RecurFibonacci (size_t index);


/* Description:
 * Recives an index in a fibonacci sequence and returns its value using
 * iteration. The sequence starts 0 1 1 2...
 *
 *
 * Params:
 *        index - an index in a fibonacci sequence
 *
 * Return Value: value of a givin index.
 */
size_t IterFibonacci (size_t index);


/* Description:
 * Using recursion, calculates the length of the string pointed to by s,
 * excluding the terminating null byte ('\0').
 * 
 * 
 * Params:
 *        s − This is the string whose length is to be found.
 *
 * Return Value: This function returns the length of string.
 */
 size_t Strlen(const char *s);


/* Description:
 * Using recursion, compares the two strings s1 and s2.  It returns an integer
 * less than, equal to, or greater than zero if s1 is found, respectively, to
 * be less than, to match, or be greater than s2.
 *
 *
 * Params:
 *        s1 − This is the first string to be compared
 *        s2 − This is the second string to be compared.
 *
 * Return Value: An integer less than, equal to, or greater than zero if s1 
 *        (or the first  n  bytes thereof) is found, respectively, to be less 
 *        than, to match, or be greater than s2.
 */
int Strcmp(const char *s1, const char *s2);


/* Description:
 * Using recursion copies the string pointed to by src, including the 
 * terminating null byte ('\0'), to the buffer pointed to by dest.  The strings
 * may not overlap, and the destination string dest must be large enough to 
 * receive the copy.  Beware of buffer overruns!
 *
 *
 * Params:
 *        dest − This is the pointer to the destination array where the content
 *        is to be copied.
 *        src − This is the string to be copied
 *
 * Return Value: A pointer to the destination string dest.
*/
char *Strcpy(char *dest, const char *src);


/* Description:
 * Using recursion appends the src string to the dest string, overwriting the
 * terminating null byte ('\0') at the end of dest, and then adds a terminating
 * null byte. The strings may not overlap, and  the  dest  string  must  have 
 * enough space for the result.  If dest is not large enough, program behavior
 * is unpredictable.
 * 
 * 
 * Params:
 *        dest − This is pointer to the destination array, which should contain
 *        a C string, and should be large enough to contain the concatenated
 *        resulting string.
 *        src − This is the string to be appended. This should not overlap 
 *        the destination.
 *
 * Return Value: This function returns a pointer to the resulting string dest.
 */       
char *Strcat(char *dest, const char *src);



/* Using recursion finds the first occurrence of the substring needle in
 * the string haystack. The terminating null bytes ('\0') are not compared.     
 * 
 * 
 * Params:
 *        haystack − This is the main C string to be scanned
 * 
 *        needle − This is the small string to be searched with-in haystack
 *        string.
 *
 *        Return Value: These functions return a pointer to the beginning of the
 *               located substring, or NULL if the substring is not found.
 */  
char *Strstr(const char *haystack, const char *needle);


/* Using recursion reverse the direction of the singly linked list.
 *
 * 
 * Params: 
 *        node - reference node.
 * Return value : 
 *        a pointer to the new first node of the list.
*/
sl_node_t *RecurSLLFlip(sl_node_t *node);




#endif /* _REC_H_  */