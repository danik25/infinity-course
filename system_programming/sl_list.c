/****************************************************************
*                Data Structures - Singly Linked List           *
* Related files:                                                *
* #Version: V 1                                                 *
* writer: Dani Benjamin                                         *
* reviewer: Oded Bryt                                           *
* last update: 10.4.19                                          *
****************************************************************/ 

#include <stdlib.h> /* malloc */
#include <assert.h> /* assert */
#include "sl_list.h"


/****************************************************************/

sl_node_t *SLLCreate(void *data, sl_node_t *next)
{
    sl_node_t *sl_node = NULL;
    sl_node = (sl_node_t*) malloc(sizeof(sl_node_t));
    if (NULL == sl_node)
    {
        return NULL;
    }
    
    sl_node->data = data;
    sl_node->next = next;
    
    return (sl_node); 
}

/****************************************************************/

void SLLDestroy(sl_node_t *node)
{
    sl_node_t *tmp = NULL;
    
    while(node != NULL)
    {
        tmp = node;
        node = node->next;
        free(tmp);
    }
}

/****************************************************************/

sl_node_t *SLLInsert(sl_node_t *node_to_insert, sl_node_t *node_location)
{
	void *tmp = NULL;
	
	assert(node_location);
	
	/* saving the location data */
	tmp = node_location->data;
	
	/* switching the data between the inserted node and the location node */
	node_location->data = node_to_insert->data;
	node_to_insert->data = tmp;
	
	node_to_insert->next = node_location->next;
	node_location->next = node_to_insert;

	return (node_location);
}

/****************************************************************/

sl_node_t *SLLInsertAfter(sl_node_t *node_to_insert, sl_node_t *node_location)
{
    assert(node_to_insert);
    assert(node_location);
    
    node_to_insert->next = node_location->next;
    node_location->next = node_to_insert;
    
    return (node_to_insert); 
}

/****************************************************************/

sl_node_t *SLLRemove(sl_node_t *node_to_remove)
{
	sl_node_t *tmp_next = NULL;
	void *tmp = NULL;
	
	assert(node_to_remove);
	
	if(NULL == node_to_remove->next)
	{
		return (NULL);
	}
	
	/* switching the data between the removed node and the next one */
	tmp = node_to_remove->next->data;
	node_to_remove->next->data = node_to_remove->data;
	node_to_remove->data = tmp;
	
	/* detaching the pointer to the removed node */
	tmp_next = node_to_remove->next;
	node_to_remove->next = node_to_remove->next->next;
	tmp_next->next = NULL;
	
	return (tmp_next);
}

/****************************************************************/

sl_node_t *SLLRemoveAfter(sl_node_t *location)
{
	sl_node_t *tmp = NULL;
	
	assert(location);
	
	if(NULL == location->next)
	{
		return NULL;
	}

    tmp = location->next;
    location->next = location->next->next; 

    tmp->next = NULL;
    
    return (tmp);  
}
/****************************************************************/

int SLLForEach(sl_node_t *node, action_function_t func, void *param)
{
    int status = 0;
    
    assert(node);
    assert(func);
    
    while(node != NULL)
    {
        status = func(node->data, param);
        if(status)
        {
            return (status); /*fail*/
        }
        node = node->next;
    }
    return (status);
}

/****************************************************************/

sl_node_t *SLLFind(sl_node_t *node, is_match_function_t func, void *param)
{
    assert(func);
    while(node != NULL)
    {
        if(func(node->data, param))
        {
            return (node);
        }
        node = node->next;
    }
    return (NULL);
}

/****************************************************************/

size_t SLLCount(const sl_node_t *node)
{
	size_t cnt = 0;
	
	if(NULL == node)
	{
		return (0);
	}
	
	while(node != NULL)
	{
		node = node->next;
		++cnt;
	}
	
	return (cnt);
}

/****************************************************************/
sl_node_t *SLLFlip(sl_node_t *node)
{
    sl_node_t *tmp = NULL;
	sl_node_t *tmp_next = NULL;
	
	assert(node);
	
	while(node)
	{
	    /* appending pointers */
	    tmp = tmp_next;
	    tmp_next = node;
	    node = node->next;
	    
	    /* assign "next" to "previous" */
	    tmp_next->next = tmp;
	}
	
	node = tmp_next;
	
	return (node);
}

/****************************************************************/

sl_node_t *SLLFindIntersection(const sl_node_t *node_1, 
       						   const sl_node_t *node_2)
{
	int gap = 0;
	
	assert(node_1);
	assert(node_2);
	
	gap = SLLCount(node_1) - SLLCount(node_2);
	if(gap > 0)
	{
		while(gap > 0)
		{
			node_1 = node_1->next;
			--gap;
		}
	}
	else
	{
		gap *= -1;
		while(gap > 0)
		{
			node_2 = node_2->next;
			--gap;
		}
	}
	while(node_1 != NULL)
	{
		if(node_1 == node_2)
		{
			return ((sl_node_t *)node_1);
		}
		
		node_1 = node_1->next;
		node_2 = node_2->next;
	}
	return (NULL);	
}

/****************************************************************/

int SLLHasLoop(sl_node_t *node)
{
	sl_node_t *by_one = NULL;
	sl_node_t *by_two = NULL;
	
	assert(node);
	
	by_one = node;
	by_two = node->next;
	
	while (by_one && by_two && by_two->next)
	{
		if(by_one == by_two)
		{
			return (1);
		}
		by_one = by_one->next;
		by_two = by_two->next->next;
	}
	return (0);
}


