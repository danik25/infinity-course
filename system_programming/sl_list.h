/****************************************************************
*                Data Structures - Singly Linked List           *
* Related files:                                                *
* #Version: V 1                                                 *
* writer: Dani Benjamin                                         *
* reviewer: Oded Bryt                                           *
* last update: 10.4.19                                          *
****************************************************************/

#ifndef _SL_LIST_
#define _SL_LIST_
#include <stddef.h> /* size_t */
 
/*
 * A singly linked list is a data structure consisting of a group of nodes,
 * linked together. Each node consists of data elements and a link
 * to the next node. The main advantage of linked lists over simple arrays 
 * is that the nodes can be inserted anywhere in the list without reallocating
 * the entire array because the nodes in a linked list do not need to be stored
 * contiguously in memory. However, nodes in the list cannot be accessed by 
 * index but must be accessed by traversing the list.
*/  

/* API Design:
	
		This API design is an infrastructure module, allowing the user
		to access the structure.
		With the ability to access the structure, the user can add nodes,
		delete nodes, modify the structure itself as well as the data in
		any of it's fields.

		This design supports any data type. 
	

	******************     *****************    *****************    
  	* DATA  *   next * --> *  DATA  * next * -> *  DATA  * next *---> NULL 
	******************     *****************    *****************
	   	  1st node 		    2nd node  ....n'th node...    Last node	
*/


/* typedef on function pointer for readability - both functions are given */
/* as arguments to SLLForEach and SLLFind respectively. */
typedef int (action_function_t)(void *data, void* param);
typedef int (is_match_function_t)(void *data, void* param);

 
typedef struct sl_node {
		    	  	
		void * data;
		struct sl_node *next; 

} sl_node_t;



/* Create a node, can connect the new node to other existing nodes.

 * Params:
		  next - pointer to the node that will be pointed by the created node.
				 (NULL if node does not point to any other node).
		  data - pointer to the data that will be inserted to the node.
 
 * Return Value: pointer to the created node. 
*/

sl_node_t *SLLCreate(void *data, sl_node_t *next);



/* Destroy the linked list - free all allocated memory from a given position.

 * Params: 
		  node - reference node.

 * Time complexity: O(n).
*/
void SLLDestroy(sl_node_t *node);	  



/* Insert a node before a given location.

 * Params: 
		  node_to_insert - desired node to be inserted. 
		  node_location  - node that will be pointed to by the inserted node.
		  

 * Return value : pointer to inserted node
*/

sl_node_t *SLLInsert(sl_node_t *node_to_insert, sl_node_t *node_location);



/* Insert a node after a given location.

 * Params: 
		  node_location  - the node that will point to the inserted node.
		  node_to_insert - the node to be inserted.

 * Return value : pointer to inserted node. 
*/

sl_node_t *SLLInsertAfter(sl_node_t *node_to_insert, sl_node_t *node_location);



/* Remove a node from a given location.Cannot remove the last node from a list.

 * Params: 
		  node_to_remove - the node that will be removed from list.

 * Return value : a pointer to the removed node.
*/
sl_node_t *SLLRemove(sl_node_t *node_to_remove);


/* Remove a node from the list after a given location.

 * Params: 
		  location  - the node that points to the node to be removed.

 * Return value : a pointer to the removed node. 
*/
sl_node_t *SLLRemoveAfter(sl_node_t *location);


/* Apply <action_func> on all nodes starting from <node>.

 * Params: 
		  node - reference node.

 * Return value : success / type of failure.
*/
int SLLForEach(sl_node_t *node, action_function_t func, void *param);



/* Find if a certain data matches data on the list.

 * Params: 
		  node - reference node.
		  data - the data to be found.	

 * Return value : pointer to node that holds relevant data.
*/
sl_node_t *SLLFind(sl_node_t *node, is_match_function_t func, void *param);


/* Count the number of nodes on the list, from a reference node.

 * Params: 
		  node - reference node.

 * Return value : number of nodes on the list.
*/
size_t SLLCount(const sl_node_t *node);



/* Reverse the direction of the singly linked list.

 * Params: 
		  node - reference node.

 * Return value : a pointer to the new first node of the list.
*/
sl_node_t *SLLFlip(sl_node_t *node);


/* Find intersections - a shared node between two lists.

 * Params: 
		  node_1 - reference node of first list.
  		  node_2 - reference node of second list.

 * Time complexity: O(n).

 * Return value : a pointer to the shared node.
*/
sl_node_t *SLLFindIntersection(const sl_node_t *node_1, 
       						   const sl_node_t *node_2);



/* Finds if a loop exists within the list.

 * Params: 
		  node  - reference node.
	
 * Return value : 1 if loop found, 0 if not.
*/
int SLLHasLoop(sl_node_t *node);


#endif /* _SL_LIST_ */
