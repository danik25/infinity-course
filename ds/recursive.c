/****************************************************************
*                                 Recursion 
*                       ==============================
* File Name: recursive.h
* Related files: recursive.c, recursive_test.c
* #Version: V 1.8
* Writer: Dani Benjamin                         
* Reviewer: Ayala Rosel 
* Last update: 28/05/19
****************************************************************/
#include <string.h> /* strncmp */
#include <assert.h> /* assert */

#include "sl_list.h"
#include "recursive.h"

size_t RecurFibonacci (size_t index)
{
    if(index == 0 || index == 1)
    {
        return (index);
    }
    return (RecurFibonacci(index - 1) + RecurFibonacci(index - 2));
}

size_t IterFibonacci (size_t index)
{
    size_t first = 0;
    size_t second = 1;
    size_t i = 0;
    size_t ans = 1;
    
    if(index == 0 || index == 1)
    {
        return(index);
    }
    for(i = 1; i < index; ++i)
    {
        ans = first + second;
        first = second;
        second = ans;
    }

    return(ans);
}

 size_t Strlen(const char *s)
 {
    assert(s);

    if('\0' == *s)
    {
        return(0);
    }
    return(1 + Strlen(++s));
 }

int Strcmp(const char *s1, const char *s2)
{
    assert(s1);
    assert(s2);

    if('\0' == *s1 || *s1 != *s2)
    {
        return(*s1 - *s2);
    }

    return(Strcmp(++s1, ++s2));
}

char *Strcpy(char *dest, const char *src)
{
    assert(dest);
    assert(src);

    *dest = *src;
    if('\0' != *src)
    {
        Strcpy(dest + 1, src + 1);
    }

    return(dest);
}

char *Strcat(char *dest, const char *src)
{
    Strcpy(dest + strlen(dest), src);
    return(dest);
}

char *Strstr(const char *haystack, const char *needle)
{
    if(0 == strncmp(haystack, needle, strlen(needle)))
    {
        return((char*)haystack);
    }
    if('\0' == *haystack)
    {
        return(NULL);
    }
    return(Strstr(haystack + 1, needle));
}

sl_node_t *RecurSLLFlip(sl_node_t *node)
{
    sl_node_t *head = NULL;

    assert(node);

    if (NULL == node->next) 
    {
       return(node);    
    }

    head = RecurSLLFlip(node->next); 
    node->next->next  = node;  
    node->next  = NULL;          
  
    return(head);
}