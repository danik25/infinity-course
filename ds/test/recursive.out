/home/xyz/git/dani-benjamin/ds/recursive.out